awx/api/serializers.py

```diff
@@ -4630,41 +4610,49 @@ def validate(self, attrs):
         for job in attrs['jobs']:
-            if 'credentials' in job:
-                [requested_use_credentials.add(cred) for cred in job['credentials']]
-            if 'labels' in job:
-                [requested_use_labels.add(label) for label in job['labels']]
-            if 'instance_groups' in job:
-                [requested_use_instance_groups.add(instance_group) for instance_group in job['instance_groups']]
+            for cred in job.get('credentials', []):
+                requested_use_credentials.add(cred)
+            for label in job.get('labels', []):
+                requested_use_labels.add(label)
+            for instance_group in job.get('instance_groups', []):
+                requested_use_instance_groups.add(instance_group)
 
-        self.check_organization_permission(attrs, request)
-        self.check_unified_job_permission(request, requested_ujts)
-        if requested_use_inventories or 'inventory' in attrs:
-            self.check_inventory_permission(attrs, request, requested_use_inventories)
+        key_to_obj_map = {
+            "unified_job_template": {obj.id: obj for obj in UnifiedJobTemplate.objects.filter(id__in=requested_ujts)},
+            "inventory": {obj.id: obj for obj in Inventory.objects.filter(id__in=requested_use_inventories)},
+            "credentials": {obj.id: obj for obj in Credential.objects.filter(id__in=requested_use_credentials)},
+            "labels": {obj.id: obj for obj in Label.objects.filter(id__in=requested_use_labels)},
+            "instance_groups": {obj.id: obj for obj in InstanceGroup.objects.filter(id__in=requested_use_instance_groups)},
+            "execution_environment": {obj.id: obj for obj in ExecutionEnvironment.objects.filter(id__in=requested_use_execution_environments)},
+        }
 
-        if requested_use_credentials:
-            self.check_credential_permission(request, requested_use_credentials)
+        ujts = {}
+        for ujt in key_to_obj_map['unified_job_template'].values():
+            ujts.setdefault(type(ujt), [])
+            ujts[type(ujt)].append(ujt)
 
-        if requested_use_labels:
-            self.check_label_permission(request, requested_use_labels)
+        unallowed_types = set(ujts.keys()) - set([JobTemplate, Project, InventorySource, WorkflowJobTemplate])
+        if unallowed_types:
+            type_names = ' '.join([cls._meta.verbose_name.title() for cls in unallowed_types])
+            raise serializers.ValidationError(_("Template types {type_names} not allowed in bulk jobs").format(type_names=type_names))
 
-        if requested_use_instance_groups:
-            self.check_instance_group_permission(request, requested_use_instance_groups)
+        for model, obj_list in ujts.items():
+            role_field = 'execute_role' if isinstance(model, (JobTemplate, WorkflowJobTemplate)) else 'update_role'
+            self.check_list_permission(model, set([obj.id for obj in obj_list]), role_field)
```

Comment:
not sure if this is working. But I have failing functional tests, which were passing before.
```
FAILED test_bulk.py::test_bulk_job_launch_no_access_to_job_template - AssertionError: Response data: OrderedDict([('id', '1'), ('type', 'workflow_job'), ('url', '/api/v2/workflow_jobs/1/'), ('related', "OrderedDict([('created_by', '/ap...
FAILED test_bulk.py::test_bulk_job_launch_inventory_no_access - AssertionError: Response data: OrderedDict([('id', '1'), ('type', 'workflow_job'), ('url', '/api/v2/workflow_jobs/1/'), ('related', "OrderedDict([('created_by', '/api/v2/u...
================================================================================================= 2 failed, 10 passed, 2 warnings in 13.90s ==================================================================================================
```